feature: send adoption register notification

eTaphee opened the pull request:
<!-- 필요한 경우 팀원들과 의논하여 양식을 변경할 수 있습니다. -->
<!-- PR은 상세히 적어주시는게 좋습니다. -->
<!-- 이해를 돕기위한 이미지를 써도 좋습니다. -->

## Changes
<!-- 어떤점이 변경되었는지 적어주세요. -->
- 분양 게시글이 등록 시 `AdoptionEvent` 이벤트를 등록합니다.
- 등록된 이벤트는 `adoption.created` 라우팅 키에 의해 `queue.notify.fcm.adoption` 큐로 바인딩됩니다.
- `FcmAdoptionConsumer`가 큐를 구독하여 메시지를 처리합니다.
  - `topic.shelter.{id}` 토픽을 기준으로 fcm 메시지를 전송합니다.
  - 전송 후 해당 토픽을 구독한 사용자 정보를 가져와 `Notification`, `NotificationReceiver` 엔티티를 저장합니다.

![adoption-process drawio](https://github.com/banzzokee/banzzokee-back/assets/43163409/62b2da0f-a919-4b6c-9dd2-1b28b5033da0)


## Background
<!-- 이 PR이 진행된 배경을 적어주세요. -->
팔로우한 보호소가 분양 게시글을 등록하면, 사용자는 알림을 받습니다.

## Discuss
<!-- 토론할 내용이 있다면 적어주세요. -->

## Issues
<!-- 관련된 이슈를 #{issueNumber}를 통해 태그해주세요. -->
<!-- ex) 해결한 이슈: #1, #2 -->
<!-- ex) 구현이 필요한 이슈: #3, #4 -->
해결한 이슈: #134 
구현이 필요한 이슈: #81 

## Execute
<!-- 실행된 결과에 대해 적어주세요. -->
<!-- 어떻게 테스트를 진행했는지, 어떤 검토를 거쳤는지 적어주시면 좋습니다. -->
<!-- 테스트 코드는 중요합니다! -->
- Adoption 등록 시 이벤트 발행 코드가 호출되는지 검증하는 코드를 추가했습니다.
- 메시지 처리 시 분양 게시글이 존재하지 않으면 `AdoptionNotFoundException` 를 발생합니다.
- 메시지 처리 성공 시 FCM 메시지 전송 호출과 `Notification`, `NotificationReceiver` 엔티티 저장을 검증합니다.

## Reference
<!-- 참조한 레퍼런스가 있다면 적어주세요. -->

eTaphee committed:
feature: 분양 게시글 등록 알림 구현

- 분양 게시글 등록 시 메시지 발행
- 메시지 처리에서 FCM 메시지 전송 및 결과 저장

eTaphee committed:
test: 분양 게시글 등록 처리 테스트 작성

- 게시글 등록 시 publishEvent() 호출 검증
- 메시지 처리 시 메시지 전송 호출 검증
- 메시지 처리 시 알림 데이터 저장 검증

eTaphee committed:
feature: 북마크 등록, 삭제 시 토픽 이벤트 메시지 등록 구현

eTaphee committed:
test: 북마크 등록, 삭제 시 이벤트 메시지 등록 호출 검증

eTaphee committed:
feature: 분양 게시글 상태 변경 알림 구현

- 분양 게시글 상태 변경 시 메시지 발행
- 메시지 처리에서 FCM 메시지 전송 및 결과 저장

nadoran78 commented on src/main/java/homes/banzzokee/domain/adoption/entity/convertor/ImagesConvertor.java:
```diff
@@ -16,15 +17,20 @@ public class ImagesConvertor implements AttributeConverter<List<S3Object>, Strin
+    if (attribute == null || attribute.isEmpty()) {
+      return null;
+    }
 
```
아 제가 null 처리를 안 했었네요 감사합니다.

eTaphee committed:
test: 분양 게시글 상태 변경 테스트 추가, 상태 변경 처리 테스트 작성

- 게시글 상태 변경 시 publishEvent() 호출 검증
- 메시지 처리 시 메시지 전송 호출 검증
- 메시지 처리 시 알림 데이터 저장 검증

nadoran78 reviewed (approved):
고생 많으셨습니다!

eTaphee committed:
feature: 리뷰 게시글 등록 알림 구현

- 리뷰 등록 시 메시지 발행
- 메시지 처리에서 FCM 메시지 전송 및 결과 저장

eTaphee committed:
test: 리뷰 알림 테스트 작성

- 리뷰 등록 시 이벤트 메시지 등록 호출 검증
- 리뷰 이벤트 처리 검증
- 리뷰 이벤트 실패 검증

eTaphee committed:
Merge pull request #143 from banzzokee/feature/send-review-register-notification

feature: send review register notification

eTaphee committed:
Merge pull request #139 from banzzokee/feature/send-adoption-status-changed-notification

feature: send adoption status changed notification

eTaphee committed:
chore: 충돌 해결
